sw/device/lib/dif/dif_hmac.c

```diff
@@ -0,0 +1,208 @@
+      config->message_endianness < kDifHmacBigEndian ||
+      config->message_endianness > kDifHmacLittleEndian ||
+      config->mode < kDifHmacModeHmac || config->mode > kDifHmacModeSha256) {
+    return kDifHmacInvalidArgument;
+  }
+
+  // Clear the config, stopping the SHA engine.
+  mmio_region_write32(config->base_addr, HMAC_CFG_REG_OFFSET, 0);
+
+  // Clear INTER_STATE.
+  mmio_region_write32(config->base_addr, HMAC_INTR_STATE_REG_OFFSET,
+                      (1 << HMAC_INTR_ENABLE_HMAC_DONE) |
+                          (1 << HMAC_INTR_STATE_FIFO_FULL) |
+                          (1 << HMAC_INTR_STATE_HMAC_ERR));
+
+  // Set the byte-order of the input message.
+  uint32_t device_config = 0;
+  if (config->message_endianness == kDifHmacBigEndian) {
+    device_config |= (1 << HMAC_CFG_ENDIAN_SWAP);
+  }
+  // Set the byte-order of the digest.
+  if (config->digest_endianness == kDifHmacBigEndian) {
+    device_config |= (1 << HMAC_CFG_DIGEST_SWAP);
+  }
+
+  // Set HMAC engine to HMAC or SHA256-only mode.
+  if (config->mode == kDifHmacModeHmac) {
+    device_config |= (1 << HMAC_CFG_HMAC_EN);
+
+    // Set the key.
+    for (int i = 0; i < HMAC_KEY_REG_COUNT; i++) {
+      ptrdiff_t offset = i * sizeof(uint32_t);
+      mmio_region_write32(config->base_addr, HMAC_KEY0_REG_OFFSET + offset,
+                          config->key[i]);
+    }
+  }
+  device_config |= (1 << HMAC_CFG_SHA_EN);
+
+  // Write the configuration.
+  mmio_region_write32(config->base_addr, HMAC_CFG_REG_OFFSET, device_config);
+
+  // Begin SHA or HMAC operation.
+  mmio_region_nonatomic_set_bit32(config->base_addr, HMAC_CMD_REG_OFFSET,
+                                  HMAC_CMD_HASH_START);
+
+  // Setup state tracking.
+  hmac->base_addr = config->base_addr;
+
+  return kDifHmacNoError;
+}
+
+dif_hmac_error_t dif_hmac_push_fifo(const dif_hmac_t *hmac, void *data,
+                                    size_t len, size_t *bytes_sent) {
+  if (!hmac || !data) {
+    return kDifHmacInvalidArgument;
+  }
+
+  const uint8_t *data_sent = (uint8_t *)data;
+  int words_remaining = len / sizeof(uint32_t);
+  int bytes_remaining = len % sizeof(uint32_t);
+
+  // Write whole words until the FIFO fills or there are no words left to send.
+  while (words_remaining > 0 && get_fifo_available_space(hmac) > 0) {
+    uint32_t word;
+
+    memcpy(&word, data_sent, sizeof(uint32_t));
+    mmio_region_write32(hmac->base_addr, HMAC_MSG_FIFO_REG_OFFSET, word);
+    data_sent += sizeof(uint32_t);
+    words_remaining--;
+  }
+
+  // Once all words are written write remaining bytes individually.
+  while (words_remaining == 0 && bytes_remaining > 0 &&
+         get_fifo_available_space(hmac) > 0) {
+    mmio_region_write8(hmac->base_addr, HMAC_MSG_FIFO_REG_OFFSET, *data_sent);
+    data_sent += sizeof(uint8_t);
+    bytes_remaining--;
+  }
+
+  if (bytes_sent) {
+    *bytes_sent = len - (words_remaining * sizeof(uint32_t)) - bytes_remaining;
+  }
+
+  return kDifHmacNoError;
+}
```

Comment:
The semantics here are that this function will (except for invalid arguments) always return `kDifHmacNoError`, even if no bytes were sent.  I'm thinking about if those are the correct semantics.  I could envision a caller doing:

```
char *buffer = ...;
size_t buffer_size = N;

size_t written = 0;
while (written < buffer_size) {
  size_t bytes_sent;
  const dif_hmac_error_t err = dif_hmac_fifo_push(hmac, buffer + written, buffer_size - written, &bytes_sent);
  if (err != kDifHmacNoError) {
    // Error reporting.
    break;
  }
}
```
Maybe I'm being paranoid here.  If the HMAC is running then it shouldn't stay full forever.  I guess if a caller tries to add data and the HMAC isn't running then this could block forever.

I think this function should return an error when it can write no data.  My justification: The function is `dif_hmac_fifo_push`, and if it pushed no data then it did not do the thing it's named.

Reply:
I like this, especially with the argument that these functions should be getting their own error types. In practice the copy process seems slow enough that the FIFO never actually fills up under this implementation, so it makes even more sense to return an error if the buffer isn't consumed in full.